### PR TITLE
Bitcoin integration

### PR DESCRIPTION
**Pull request by trenatov**
Bitcoin integration #41

### CODE REVIEW

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    override suspend fun getBlock(blockNumber: Int): UnifiedBlock {
+        val blockHash = bitcoinRpcClient.getBlockHash(blockNumber)
+        val block = bitcoinRpcClient.getBlock(blockHash)
+
+        return toUnifiedBlock(block)
+    }
+
+    private suspend fun toUnifiedBlock(btcBlock: BitcoinBlock): UnifiedBlock {
+        return UnifiedBlock(
+                toUnifiedTransactions(btcBlock.transactions),
```
Comment (layko-andrey) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: May be in milliseconds?

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                btcBlock.time.toLocalDateTimeInSeconds(),
+                btcBlock.height,
+                btcBlock.hash
+        )
+    }
+
+    private suspend fun toUnifiedTransactions(btcTransactions: List<BitcoinTransaction>): List<UnifiedTransaction> {
+        val transactions = btcTransactions.toMutableList()
+        //delete coinbase transaction (miner award)
```
Comment (layko-andrey) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: Can we detect coinbase transaction using transaction data, not by index in transaction list?

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        transactions.removeAt(0)
+
+        return transactions.flatMap { obtainTransactions(it) }
+    }
+
+    private suspend fun obtainTransactions(btcTransaction: BitcoinTransaction): List<UnifiedTransaction> {
```
Comment (layko-andrey) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: Why we do that?

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                            .get("details")
+                            .asIterable()
+                            .filter { it.get("vout").asInt() == outputNumber }
+                            .map { it.get("address").asText() }.first()
+                }
+                .awaitSingle()
+    }
+
+    private data class Command(val method: String, val params: List<Any>)
+
```
Comment (layko-andrey) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: This should be enum

File: src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt [hunk 5]
```diff
@@ -4,7 +4,7 @@ import java.math.BigDecimal
```
Comment (layko-andrey) [inline src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt:4]: I don't think that it a good idea to use the same model for different structures of data. What if next chain would require maps?

**Review by layko-andrey — DISMISSED** [review]

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt (hunk 1 above)
Comment (trenatov) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: no, in seconds

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt (hunk 2 above)
Comment (trenatov) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: coinbase transaction will be allways at 0 index

File: src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt (hunk 5 above)
Comment (trenatov) [reply src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt:4]: to some extent I agree with you, but overall structure of blockchains are same

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt (hunk 3 above)
Comment (trenatov) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: will be fixed

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt (hunk 4 above)
Comment (trenatov) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: but it expects json body, thats why it is data class. If you mean to wrap this logic into Enum, will be done

File: src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt (hunk 5 above)
Comment (trenatov) [reply src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt:4]: i think we should discuss

File: src/main/kotlin/io/openfuture/state/blockchain/EthereumBlockchain.kt [hunk 6]
```diff
@@ -33,7 +33,7 @@ class EthereumBlockchain(private val web3j: Web3j) : Blockchain() {
             .map { tx ->
                 val to = tx.to ?: findContractAddress(tx.hash)
                 val amount = Convert.fromWei(tx.value.toBigDecimal(), Convert.Unit.ETHER)
-                UnifiedTransaction(tx.hash, tx.from, to, amount)
+                UnifiedTransaction(tx.hash, setOf(tx.from), setOf(to), amount)
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/EthereumBlockchain.kt]: You should create a constructor which will accept a single value

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlock.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package io.openfuture.state.blockchain.bitcoin
+
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class BitcoinBlock(
+        val hash: String,
+        val height: Long,
+        val time: Long,
+        @field:JsonProperty("tx") val transactions: List<BitcoinTransaction>
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlock.kt]: Next line

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        btcTransaction.inputs.toMutableList().removeAt(0)
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: why do you remove the first element in here? 

BTW, you can use `drop(1)` function

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        val inputAddresses = btcTransaction.inputs
+                .toMutableList()
+                .map { bitcoinRpcClient.getInputAddress(it.txId!!, it.outputNumber!!) }
+                .toSet()
+
+        //remove if its is `change address`
+        btcTransaction.outputs.toMutableList().removeAll { it.addresses.any { address -> inputAddresses.contains(address) } }
+        return btcTransaction.outputs
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: Are you sure here is everything good with code format?

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+class BitcoinRpcClient(private val bitcoinWebClient: WebClient, private val objectMapper: ObjectMapper) {
+
+    suspend fun getLatestBlockHash(): String {
+        val command = Command(GET_LATEST_BLOCK_COMMAND, emptyList())
+        return bitcoinWebClient
+                .post()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(BodyInserters.fromValue(command))
+                .retrieve()
+                .bodyToMono(String()::class.java)
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: String()?

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+@Component
+class BitcoinRpcClient(private val bitcoinWebClient: WebClient, private val objectMapper: ObjectMapper) {
+
+    suspend fun getLatestBlockHash(): String {
+        val command = Command(GET_LATEST_BLOCK_COMMAND, emptyList())
+        return bitcoinWebClient
+                .post()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(BodyInserters.fromValue(command))
+                .retrieve()
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: You should use Kotlin version

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                            .filter { it.get("vout").asInt() == outputNumber }
+                            .map { it.get("address").asText() }.first()
+                }
+                .awaitSingle()
+    }
+
+    private data class Command(val method: String, val params: List<Any>)
+
+    private companion object {
+        private const val GET_LATEST_BLOCK_COMMAND = "getbestblockhash"
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: You should not have consts here. You should inline them

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 13]
```diff
@@ -0,0 +1,92 @@
+package io.openfuture.state.blockchain.bitcoin
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import kotlinx.coroutines.reactive.awaitSingle
+import org.springframework.http.MediaType
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.BodyInserters
+import org.springframework.web.reactive.function.client.WebClient
+
+@Component
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: Tests?

`@RestClientTests`

Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: github.com/ruslanys/ifunny

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt (hunk 10 above)
Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: .awaitSingle<BitcoinResponse<BlockHeightResponse>>()

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+    override suspend fun getBlock(blockNumber: Int): UnifiedBlock {
+        val blockHash = bitcoinRpcClient.getBlockHash(blockNumber)
+        val block = bitcoinRpcClient.getBlock(blockHash)
+
+        return toUnifiedBlock(block)
+    }
+
+    private suspend fun toUnifiedBlock(btcBlock: BitcoinBlock): UnifiedBlock {
+        return UnifiedBlock(
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: body expression

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+                btcBlock.time.toLocalDateTimeInSeconds(),
+                btcBlock.height,
+                btcBlock.hash
+        )
+    }
+
+    private suspend fun toUnifiedTransactions(btcTransactions: List<BitcoinTransaction>): List<UnifiedTransaction> {
+        //skip coinbase transaction (miner award)
+        return btcTransactions
+                .drop(1)
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: set the comment at the proper line

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    private suspend fun toUnifiedBlock(btcBlock: BitcoinBlock): UnifiedBlock {
+        return UnifiedBlock(
+                toUnifiedTransactions(btcBlock.transactions),
+                btcBlock.time.toLocalDateTimeInSeconds(),
+                btcBlock.height,
+                btcBlock.hash
+        )
+    }
+
+    private suspend fun toUnifiedTransactions(btcTransactions: List<BitcoinTransaction>): List<UnifiedTransaction> {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: transactions

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    override suspend fun getBlock(blockNumber: Int): UnifiedBlock {
+        val blockHash = bitcoinRpcClient.getBlockHash(blockNumber)
+        val block = bitcoinRpcClient.getBlock(blockHash)
+
+        return toUnifiedBlock(block)
+    }
+
+    private suspend fun toUnifiedBlock(btcBlock: BitcoinBlock): UnifiedBlock {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: block

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    private suspend fun obtainTransactions(btcTransaction: BitcoinTransaction): List<UnifiedTransaction> {
+        val inputAddresses = btcTransaction.inputs
+                .map { bitcoinRpcClient.getInputAddress(it.txId!!, it.outputNumber!!) }
+                .toSet()
+
+        return btcTransaction.outputs
+                .filter { it.addresses.isNotEmpty() }
+                .filter {
+                    //skip `change addresses`
+                    !it.addresses.any { address -> inputAddresses.contains(address) }
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: double-check

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+                }
+                .map {
+                    UnifiedTransaction(
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: one line

Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchain.kt]: max 120 symbols

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 20]
```diff
@@ -0,0 +1,77 @@
+package io.openfuture.state.blockchain.bitcoin
+
+import org.springframework.http.MediaType
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.BodyInserters
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.awaitBody
+
+@Component
+class BitcoinRpcClient(private val bitcoinWebClient: WebClient) {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: Replace `bitcoinWebClient` with WebClientBuilder and initialization in place.

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+
+    suspend fun getLatestBlockHash(): String {
+        val command = Command("getbestblockhash", emptyList())
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinRpcClient.kt]: default value

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinTransaction.kt [hunk 22]
```diff
@@ -0,0 +1,35 @@
+package io.openfuture.state.blockchain.bitcoin
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import java.math.BigDecimal
+
+data class BitcoinTransaction(
+        val hash: String,
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinTransaction.kt]: empty lines

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinTransaction.kt [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    )
+
+    data class Output(val value: BigDecimal) {
+
+        val addresses: MutableSet<String> = mutableSetOf()
+
+        @Suppress("UNCHECKED_CAST")
+        @JsonProperty("scriptPubKey")
+        private fun unpackNested(scriptPubKey: Map<String, Any>) {
+            val addresses = scriptPubKey["addresses"]
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinTransaction.kt]: rename

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinTransaction.kt [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+            addresses?.let {
+                this.addresses.addAll((it as List<String>).toMutableSet())
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinTransaction.kt]: redundant invocation

File: src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt [hunk 25]
```diff
@@ -4,7 +4,19 @@ import java.math.BigDecimal
 
 data class UnifiedTransaction(
         val hash: String,
-        val from: String,
-        val to: String,
+        val from: Set<String>,
+        val to: Set<String>,
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt]: to – single
from – plural

File: src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt [hunk 26]
```diff
@@ -4,7 +4,19 @@ import java.math.BigDecimal
...
         val amount: BigDecimal
-)
+) {
+    constructor(
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/dto/UnifiedTransaction.kt]: one line

File: src/main/kotlin/io/openfuture/state/config/BitcoinWebclientConfig.kt [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+import io.openfuture.state.property.BitcoinProperties
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.http.client.reactive.ReactorClientHttpConnector
+import org.springframework.http.client.reactive.ReactorResourceFactory
+import org.springframework.web.reactive.function.client.WebClient
+
+
+@Configuration
+class BitcoinWebclientConfig {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/config/BitcoinWebclientConfig.kt]: completely remove

File: src/main/kotlin/io/openfuture/state/property/BitcoinProperties.kt [hunk 28]
```diff
@@ -0,0 +1,16 @@
...
+import javax.validation.constraints.NotBlank
+import javax.validation.constraints.NotNull
+
+@ConstructorBinding
+@ConfigurationProperties(prefix = "bitcoin")
+@Validated
+data class BitcoinProperties(
+        @field:NotNull @field:NotBlank val nodeAddress: String?,
+        @field:NotNull @field:NotBlank val username: String?,
+        @field:NotNull @field:NotBlank val password: String?,
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/property/BitcoinProperties.kt]: does NotBlank cover NotNull verification as well?

File: src/main/kotlin/io/openfuture/state/util/DateUtils.kt [hunk 29]
```diff
@@ -13,3 +13,8 @@ fun Long.toLocalDateTime(): LocalDateTime {
 fun LocalDateTime.toEpochMilli(): Long {
     return this.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
 }
+
+fun Long.toLocalDateTimeInSeconds(): LocalDateTime {
+    return LocalDateTime.ofInstant(Instant.ofEpochSecond(this),
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/util/DateUtils.kt]: formatting

File: src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinClient.kt [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+
+import io.openfuture.state.blockchain.bitcoin.dto.*
+import org.springframework.http.MediaType
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.BodyInserters
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.awaitBody
+
+@Component
+class BitcoinClient(private val client: WebClient) {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinClient.kt]: `private val webClientBuilder: WebClient.Builder`

File: src/test/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchainTest.kt [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+import com.nhaarman.mockitokotlin2.given
+import com.nhaarman.mockitokotlin2.mock
+import io.openfuture.state.util.createDummyBitcoinBlock
+import io.openfuture.state.util.createDummyUnifiedBlock
+import kotlinx.coroutines.runBlocking
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+
+class BitcoinBlockchainTest {
```
Comment (ruslanys) [inline src/test/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchainTest.kt]: Remove this test

**Review by ruslanys — CHANGES_REQUESTED** [review]

File: src/test/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchainTest.kt (hunk 31 above)
Comment (trenatov) [reply src/test/kotlin/io/openfuture/state/blockchain/bitcoin/BitcoinBlockchainTest.kt]: done

**Review by ruslanys — APPROVED** [review]

**Pull request merged**